compio-driver/src/key.rs

```diff
@@ -1,40 +1,160 @@
+    where
+        T: Sized,
+    {
+        BufResult(self.result.take_ready().unwrap(), self.op)
+    }
+}
+
+#[cfg(windows)]
+impl<T: OpCode + ?Sized> RawOp<T> {
+    pub fn operate_blocking(&mut self) -> io::Result<usize> {
+        use std::task::Poll;
+
+        let optr = self.as_mut_ptr();
+        let op = self.as_op_pin();
+        let res = unsafe { op.operate(optr.cast()) };
+        match res {
+            Poll::Pending => unreachable!("this operation is not overlapped"),
+            Poll::Ready(res) => res,
+        }
+    }
+}
+
+unsafe fn upcast<T: OpCode>(user_data: usize) -> *mut RawOp<dyn OpCode> {
+    user_data as *mut RawOp<T> as *mut RawOp<dyn OpCode>
+}
+
 /// A typed wrapper for key of Ops submitted into driver
 #[derive(PartialEq, Eq, Hash)]
 pub struct Key<T> {
     user_data: usize,
-    _p: std::marker::PhantomData<fn(T)>,
+    _p: PhantomData<Box<T>>,
 }
 
 impl<T> Unpin for Key<T> {}
 
-impl<T> Clone for Key<T> {
-    fn clone(&self) -> Self {
-        *self
+impl<T: OpCode + 'static> Key<T> {
+    /// Create [`RawOp`] and get the [`Key`] to it.
+    pub fn new(driver: RawFd, op: T) -> Self {
+        let header = Overlapped::new(driver);
+        let raw_op = Box::new(RawOp {
+            header,
+            cancelled: false,
+            upcast_fn: upcast::<T>,
+            result: PushEntry::Pending(None),
+            op,
+        });
+        unsafe { Self::new_unchecked(Box::into_raw(raw_op) as _) }
     }
 }
 
-impl<T> Copy for Key<T> {}
-
 impl<T> Key<T> {
     /// Create a new `Key` with the given user data.
     ///
     /// # Safety
     ///
     /// Caller needs to ensure that `T` does correspond to `user_data` in driver
     /// this `Key` is created with.
-    pub unsafe fn new(user_data: usize) -> Self {
+    pub unsafe fn new_unchecked(user_data: usize) -> Self {
         Self {
             user_data,
-            _p: std::marker::PhantomData,
+            _p: PhantomData,
         }
     }
+
+    /// Get the unique user-defined data.
+    pub const fn user_data(&self) -> usize {
+        self.user_data
+    }
+
+    /// Get the inner result if it is completed.
+    pub fn into_inner(self) -> BufResult<usize, T> {
+        unsafe { Box::from_raw(self.user_data as *mut RawOp<T>) }.into_inner()
+    }
+}
+
+impl Key<()> {
+    pub(crate) unsafe fn drop_in_place(user_data: usize) {
+        let op = &*(user_data as *const RawOp<()>);
+        let ptr = (op.upcast_fn)(user_data);
+        let _ = Box::from_raw(ptr);
+    }
+
+    pub(crate) unsafe fn upcast<'a>(user_data: usize) -> &'a mut RawOp<dyn OpCode> {
+        let op = &*(user_data as *const RawOp<()>);
+        &mut *(op.upcast_fn)(user_data)
+    }
 }
 
-impl<T> std::ops::Deref for Key<T> {
-    type Target = usize;
+impl<T> Deref for Key<T> {
+    type Target = RawOp<T>;
 
     fn deref(&self) -> &Self::Target {
-        &self.user_data
+        unsafe { &*(self.user_data as *const RawOp<T>) }
+    }
+}
+
+impl<T> DerefMut for Key<T> {
```

Comment:
Do we want to make sure that op cannot be moved out (e.g. via `mem::replace`) of Key?

Reply:
Sounds fair.